src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -182,64 +203,46 @@ int MPL_gavl_tree_create(void (*free_fn) (void *), MPL_gavl_tree_t * gavl_tree)
+                                  int mode, gavl_tree_node_s ** ret_node)
 {
-    int mpl_err = MPL_SUCCESS;
-    gavl_tree_node_s *node_ptr;
-    gavl_tree_s *gavl_tree_iptr = (gavl_tree_s *) gavl_tree;
-
-    node_ptr = (gavl_tree_node_s *) MPL_malloc(sizeof(gavl_tree_node_s), MPL_MEM_OTHER);
-    if (node_ptr == NULL) {
-        mpl_err = MPL_ERR_SHM_NOMEM;
-        goto fn_fail;
-    }
-    node_ptr->parent = NULL;
-    node_ptr->left = NULL;
-    node_ptr->right = NULL;
-    node_ptr->height = 1;
-    node_ptr->addr = (uintptr_t) addr;
-    node_ptr->len = len;
-    node_ptr->val = val;
+    int cmp_ret;
+    gavl_tree_node_s *cur_node = gavl_tree_iptr->root;
 
-    if (gavl_tree_iptr->root == NULL) {
-        gavl_tree_iptr->root = node_ptr;
-    } else {
-        DECLARE_STACK(gavl_tree_node_s *, node_stack);
-        gavl_tree_node_s *cur_node = gavl_tree_iptr->root;
-        int direction;
-        do {
-            int cmp_ret = gavl_subset_cmp_func((uintptr_t) node_ptr->addr, node_ptr->len, cur_node);
-            if (cmp_ret == SEARCH_LEFT) {
-                if (cur_node->left != NULL) {
-                    STACK_PUSH(node_stack, cur_node);
-                    cur_node = cur_node->left;
-                    continue;
-                } else {
-                    direction = SEARCH_LEFT;
-                }
-            } else if (cmp_ret == SEARCH_RIGHT) {
-                if (cur_node->right != NULL) {
-                    STACK_PUSH(node_stack, cur_node);
-                    cur_node = cur_node->right;
-                    continue;
-                } else {
-                    direction = SEARCH_RIGHT;
-                }
+    do {
+        GAVL_TREE_NODE_CMP(addr, len, cur_node, mode, cmp_ret);
+        if (cmp_ret == SEARCH_LEFT) {
+            if (cur_node->left != NULL) {
+                STACK_PUSH(gavl_tree_iptr->node_stack, cur_node);
+                cur_node = cur_node->left;
+                continue;
             } else {
-                /* we cannot insert the duplicate buffer */
-                gavl_tree_iptr->gavl_free_fn((void *) node_ptr->val);
-                MPL_free(node_ptr);
-                goto fn_exit;
+                break;
+            }
+        } else if (cmp_ret == SEARCH_RIGHT) {
+            if (cur_node->right != NULL) {
+                STACK_PUSH(gavl_tree_iptr->node_stack, cur_node);
+                cur_node = cur_node->right;
+                continue;
+            } else {
+                break;
             }
+        } else {
+            /* node match */
+            break;
+        }
+    } while (1);
 
-            if (direction == SEARCH_LEFT)
-                cur_node->left = node_ptr;
-            else
-                cur_node->right = node_ptr;
-            node_ptr->parent = cur_node;
+    *ret_node = gavl_tree_iptr->cur_node = cur_node;
+    return;
+}
+
+static void gavl_tree_rebalance(gavl_tree_s * gavl_tree_iptr)
```

Comment:
Also, instead of `gavl_tree_iptr`, how about just `tree`? The components of a name also adds to the complexity. We need add prefix/suffix/longer-name to prevent ambiguity and collision. But when there isn't the concern of ambiguity or collisions, then we should avoid the complexity that serves no purposes.

Reply:
Original tree is balanced, and after insertion it becomes unbalanced, so I would prefer to call it `rebalance`
alright I can change the name.